### PR TITLE
Chore: fix and unify visibility

### PR DESCRIPTION
**Pull request by magurotuna**
Currently, there are some problems about visibility.

### 1. Items that are not required to be `pub` have public visibility

I think that from the crate consumers' point of view, the items that should be public are:

- `diagnostic::LintDiagnostic`
- `diagnostic::Location`
- `linter::Context`
- `linter::IgnoreDirective`
- `linter::Linter`
- `linter::LinterBuilder`
- `rules::LintRule`
- `rules::get_all_rules`
- `rules::get_recommended_rules`
- `rules::<rule-name>::<rule-struct>`
- `swc_util::SwcDiagnosticBuffer`

Therefore I have made the other items private.

### 2. Different visitors have different visibility for their constructors (some are `pub`, the others are private).

The visitors don't have to be public to crate consumers because they are the implementation detail of the rules and used by `lint_module` internally. So I have made all the items about `visitor` private.

### CODE REVIEW

**Comment by magurotuna** [comment]
Deno is using its own `get_default_ts_config` function rather than `deno_lint::swc_util::get_default_ts_config`.
Is deno_lint's one being exported only for `examples/dlint`?

**Comment by bartlomieju** [comment]
> 1. Items that are not required to be pub have public visibility

Sounds good

> 2. Different visitors have different visibility for their constructors (some are pub, the others are private).

Also sounds good

> Deno is using its own get_default_ts_config function rather than deno_lint::swc_util::get_default_ts_config.
Is deno_lint's one being exported only for examples/dlint?

Yes, please keep it as is, it will be soon refactored after `AstParser` is moved to SWC.

I'm for removing unnecessary `pub` modifiers, but we don't need to change files like `scopes.rs` - they are not exported from library so they can keep `pub` modifiers on the methods - they'll only be available in the crate code itself.

File: src/swc_util.rs [hunk 1]
```diff
@@ -35,7 +35,7 @@ use swc_ecmascript::{
 };
 
 #[allow(unused)]
-pub fn get_default_es_config() -> Syntax {
+pub(crate) fn get_default_es_config() -> Syntax {
```
Comment (bartlomieju) [inline src/swc_util.rs:38]: Revert

File: src/swc_util.rs [hunk 2]
```diff
@@ -51,8 +51,7 @@ pub fn get_default_es_config() -> Syntax {
   Syntax::Es(config)
 }
 
-#[allow(unused)]
-pub fn get_default_ts_config() -> Syntax {
+pub(crate) fn get_default_ts_config() -> Syntax {
```
Comment (bartlomieju) [inline src/swc_util.rs:54]: ditto

File: src/scopes.rs [hunk 3]
```diff
@@ -22,34 +22,36 @@ pub struct Scope {
 
 impl Scope {
   // Get all declarations with a symbol.
-  pub fn ids_with_symbol(&self, sym: &JsWord) -> Option<&Vec<Id>> {
+  #[allow(dead_code)]
```
Comment (bartlomieju) [inline src/scopes.rs:25]: Please revert changes to this file

File: src/js_regex/validator.rs [hunk 4]
```diff
@@ -92,7 +92,8 @@ fn combine_surrogate_pair(lead: i64, trail: i64) -> i64 {
 }
 
 #[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
-pub enum EcmaVersion {
+#[allow(dead_code)]
+pub(crate) enum EcmaVersion {
```
Comment (bartlomieju) [inline src/js_regex/validator.rs:96]: revert

**Review by bartlomieju — CHANGES_REQUESTED** [review]

**Comment by magurotuna** [comment]
@bartlomieju 
Thanks a lot for the review and answering my question!
I fixed the part I was overdoing. Could you please check it?

**Review by bartlomieju — APPROVED** [review]
Cool! Thanks @magurotuna

**Pull request merged**